rust/ffi/node/src/table.rs

```diff
@@ -0,0 +1,221 @@
+        let write_mode_map: HashMap<&str, WriteMode> = HashMap::from([
+            ("create", WriteMode::Create),
+            ("append", WriteMode::Append),
+            ("overwrite", WriteMode::Overwrite),
+        ]);
+
+        let js_table = cx.this().downcast_or_throw::<JsBox<JsTable>, _>(&mut cx)?;
+        let buffer = cx.argument::<JsBuffer>(0)?;
+        let write_mode = cx.argument::<JsString>(1)?.value(&mut cx);
+
+        let batches = arrow_buffer_to_record_batch(buffer.as_slice(&mut cx)).or_throw(&mut cx)?;
+        let schema = batches[0].schema();
+
+        let rt = runtime(&mut cx)?;
+
+        let (deferred, promise) = cx.promise();
+        let write_mode = write_mode_map.get(write_mode.as_str()).cloned();
+
+        let aws_creds = match get_aws_creds(&mut cx, 2) {
+            Ok(creds) => creds,
+            Err(err) => return err,
+        };
+
+        let params = WriteParams {
+            store_params: Some(ObjectStoreParams {
+                aws_credentials: aws_creds,
+                ..ObjectStoreParams::default()
+            }),
+            mode: write_mode.unwrap_or(WriteMode::Append),
+            ..WriteParams::default()
+        };
+
+        js_table
+            .send(deferred, move |table, channel, deferred| {
+                rt.block_on(async move {
+                    let batch_reader = RecordBatchIterator::new(batches.into_iter().map(Ok), schema);
+                    let add_result = table.add(batch_reader, Some(params)).await;
+
+                    deferred.settle_with(&channel, move |mut cx| {
+                        let _added = add_result.or_throw(&mut cx)?;
+                        Ok(cx.boolean(true))
+                    });
+                });
+            })
+            .or_throw(&mut cx)?;
+        Ok(promise)
+    }
+
+    pub(crate) fn js_count_rows(mut cx: FunctionContext) -> JsResult<JsPromise> {
+        let js_table = cx.this().downcast_or_throw::<JsBox<JsTable>, _>(&mut cx)?;
+        let rt = runtime(&mut cx)?;
+
+        let (deferred, promise) = cx.promise();
+
+        js_table
+            .send(deferred, move |table, channel, deferred| {
+                rt.block_on(async move {
+                    let num_rows_result = table.count_rows().await;
+
+                    deferred.settle_with(&channel, move |mut cx| {
+                        let num_rows = num_rows_result.or_throw(&mut cx)?;
+                        Ok(cx.number(num_rows as f64))
+                    });
+                });
+            })
+            .or_throw(&mut cx)?;
+        Ok(promise)
+    }
+
+    pub(crate) fn js_delete(mut cx: FunctionContext) -> JsResult<JsPromise> {
+        let js_table = cx.this().downcast_or_throw::<JsBox<JsTable>, _>(&mut cx)?;
+        let rt = runtime(&mut cx)?;
+        let (deferred, promise) = cx.promise();
+        let predicate = cx.argument::<JsString>(0)?.value(&mut cx);
+
+        js_table
+            .send(deferred, move |table, channel, deferred| {
```

Comment:
Food for thought: Can we eliminate more of this boilerplate? It seems like all that is essential is the argument extraction and the call to `.delete()`. So it would be nice if the body was just something like:

```rust
let predicate: String = cx.argument::<JsString>(0)?.value(&mut cx);
extract_table(cx).call(move |table| table.delete(&predicate))
```